### ISSUE TITLE
Quantization problems, warnings in initialization and prompting of CheXagent 2b

### ISSUE THREAD

**Issue by dscarmo** [Stanford-AIMI/CheXagent#16]
Hello! We have been testing CheXagent 8B with great success, and even were able to fit it in a 8GB GPU with 4 bit quantization. 

We are very interested in migrating to CheXagent 3B with 8 bit quantization, however we have run into some issues when trying to use different data types, and some concerning warning messages.

Starting from the provided code snippet (which is very different from the CheXagent 8B one) I have the following questions: 

1. When loading the models, we get many messages of this type:

> UserWarning: for vision_model.head.mlp.fc2.bias: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)

2. When running inference on the CPU, even thought inference works, we get: 

> The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
> Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.

For both warning, Is this expected? Should it be ignored? Is there a way to suppress this warning or fix it? It is specially concerning if pad/mask tokens are possibly being set incorrectly.  

3. Finally, and probably the most important issue, we are not able when using the provided snippet, to employ 8 bit quantization, which is necessary for our use case. I have made it work by changing the CheXagent 8B class code to use CheXagent 2B weights, but I am not sure if this is the correct way to use the new weights. So, I am asking if this implementation seems correct, and asking for your seal of approval. There are some subtle differences necessary to make quantization work:

Initialization:
> class CheXagent(object):
>     def __init__(self, dtype=torch.bfloat16, device="cuda", quantization_config={"load_in_8bit": True}):
>         # step 1: Setup constant
>         self.model_name = "StanfordAIMI/CheXagent-2-3b"
>         self.dtype = dtype
>         self.device = device
> 
>         # step 2: Load Processor and Model
>         self.tokenizer = AutoTokenizer.from_pretrained(self.model_name, trust_remote_code=True)
>         self.model = AutoModelForCausalLM.from_pretrained(self.model_name, 
>                                                           quantization_config=quantization_config,
>                                                           device_map=self.device,
>                                                           torch_dtype=self.dtype, 
>                                                           trust_remote_code=True)
>         self.processor = AutoProcessor.from_pretrained(self.model_name, trust_remote_code=True)
>         self.generation_config = GenerationConfig.from_pretrained(self.model_name, trust_remote_code=True)
> 
>         self.model.eval()
> 
>         print(f"Initialized CheXagent {self.model.dtype} model.")

Generation
>     def generate(self, images, prompt: str):
>         """
>         Custom generate mixing things from old cheXagent and new cheXagent
>         """
>         images = self.to_img(images)
> 
>         inputs = self.processor(images=images, text=f" USER: <s>{prompt} ASSISTANT: <s>{''}", return_tensors="pt").to(device=self.device)
>         with torch.no_grad():
>             output = self.model.generate(**inputs, generation_config=self.generation_config, max_new_tokens=512)[0]
>         response = self.tokenizer.decode(output, skip_special_tokens=True)
>     
>         return response

Sorry if the format is wonky, and thanks for all the nice work on CheXagent!

**Comment by jbdel** [comment]
Hello. 
The latest chexagent is 
https://huggingface.co/StanfordAIMI/CheXagent-2-3b

please follow the instructions there. cheers

**Comment by dscarmo** [comment]
> Hello. 
> The latest chexagent is 
> https://huggingface.co/StanfordAIMI/CheXagent-2-3b
> 
> please follow the instructions there. cheers

Hi! Thanks for the quick answer, i think my issue was confusing and asking too many things. As I said, those instructions work on the cpu but do not work with hugging face quantization to 8bits, which is necessary for my use case. I gave the code that makes it work with quantization.

I suspect this is the case because you are not using the Processor anymore, which causes problems with image encoding. 

Do you have a code snippet that would work with quantization_config, as in the example I provided? Thanks
